### PR TITLE
applications: nrf5340_audio: Add a deinitialise function

### PR DESCRIPTION
**Pull request by gWacey**
The data FIFO module did not have a deinitialise function, so added one, with a twister test.

### CODE REVIEW

File: lib/data_fifo/data_fifo.c [hunk 1]
```diff
@@ -167,6 +167,22 @@ int data_fifo_empty(struct data_fifo *data_fifo)
 	return 0;
 }
 
+int data_fifo_deinit(struct data_fifo *data_fifo)
+{
+	__ASSERT_NO_MSG(data_fifo != NULL);
```
Comment (koffes) [inline lib/data_fifo/data_fifo.c:172]: Recommend using ASSERT with message.

File: include/data_fifo.h [hunk 2]
```diff
@@ -150,6 +150,18 @@ int data_fifo_num_used_get(struct data_fifo *data_fifo, uint32_t *alloced_num,
  */
 int data_fifo_empty(struct data_fifo *data_fifo);
 
+/**
+ * @brief Deinitialise the data_fifo.
+ *
+ * @note The fifo is emptied first, so it is the users responsibility to release any data items it
```
Comment (koffes) [inline include/data_fifo.h:156]: What happens if the user does not release all the data items?

File: tests/lib/data_fifo/src/main.c [hunk 3]
```diff
@@ -31,6 +31,21 @@ static void internal_test_remaining_elements(struct data_fifo *data_fifo, uint32
 		      num_locked, line);
 }
 
+ZTEST(suite_data_fifo, test_data_fifo_deinit_ok)
```
Comment (koffes) [inline tests/lib/data_fifo/src/main.c:34]: Nice.

Suggest we add a bit more substantial test where we init a fifo, fill it, deinit it, init it again and fill it.

**Review by koffes — CHANGES_REQUESTED** [review]

File: include/data_fifo.h (hunk 2 above)
Comment (gWacey) [reply include/data_fifo.h:156]: What will happen to any buffers that are carried in the message?
We get a message and load it with a pointer to some data (on a slab for example). When all the messages are released what happens to this item and the slab it is on? It may be lost to the system!

File: tests/lib/data_fifo/src/main.c (hunk 3 above)
Comment (gWacey) [reply tests/lib/data_fifo/src/main.c:34]: Ok will add something.

File: lib/data_fifo/data_fifo.c (hunk 1 above)
Comment (gWacey) [reply lib/data_fifo/data_fifo.c:172]: On this, It is good to test like this, but if there is an issue in the release that would have exercised this assert but it has been optimised out, then the codes action would be undetermined and potentially difficult to debug. Would be more code but better in my option to explicitly test for NULL. Actually it shouldn't be more than a test zero and branch (with branch prediction this will be even better). An error is already returned and tested for, so the code could gracefully fail, rather than go off somewhere into memory!

**Review by andvib — APPROVED** [review]

File: include/data_fifo.h [hunk 4]
```diff
@@ -150,6 +150,18 @@ int data_fifo_num_used_get(struct data_fifo *data_fifo, uint32_t *alloced_num,
...
+ *       has queued. The internal slab and message buffer are not released.
+ *
+ * @param data_fifo Pointer to the data_fifo structure.
+ *
+ * @retval 0 if success, error otherwise.
+ */
+int data_fifo_deinit(struct data_fifo *data_fifo);
```
[the remark this answers is not in the document]
Comment (gWacey) [reply include/data_fifo.h:163]: Will correct.

File: include/data_fifo.h [hunk 5]
```diff
@@ -150,6 +150,18 @@ int data_fifo_num_used_get(struct data_fifo *data_fifo, uint32_t *alloced_num,
  */
 int data_fifo_empty(struct data_fifo *data_fifo);
 
+/**
+ * @brief Uninitialise the data_fifo.
```
Comment (greg-fer) [inline include/data_fifo.h:154]:
```suggestion
 * @brief Deinitialize data_fifo.
```
We follow American English (`z`).

File: include/data_fifo.h [hunk 6]
```diff
@@ -150,6 +150,18 @@ int data_fifo_num_used_get(struct data_fifo *data_fifo, uint32_t *alloced_num,
...
+ *
+ * @note The fifo is emptied first, so it is the users responsibility to release any data items it
```
Comment (greg-fer) [inline include/data_fifo.h:156]:
```suggestion
 * @note data_fifo is emptied first, so it is the user's responsibility to release any data items it
```

**Review by greg-fer — CHANGES_REQUESTED** [review]

**Review by greg-fer — CHANGES_REQUESTED** [review]
My comments not yet applied.
Can you rebase the PR properly, as I'm getting lost what is new and what is not new among the 120 commits and 300+ files edited? Thanks.

**Review by greg-fer — APPROVED** [review]

**Review by koffes — APPROVED** [review]
